sdk/core/core-https/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -33,11 +42,39 @@ export function bearerTokenAuthenticationPolicy(
   options: BearerTokenAuthenticationPolicyOptions
 ): PipelinePolicy {
   const { credential, scopes } = options;
-  const tokenCache: AccessTokenCache = new ExpiringAccessTokenCache();
+  const tokenCache: ExpiringAccessTokenCache = new ExpiringAccessTokenCache(
+    options.tokenRefreshBufferMs
+  );
+
+  let refreshPromise: Promise<AccessToken | null> | undefined;
+
+  let refreshAttemptsBufferMs = options.refreshAttemptsBufferMs || 30000;
+
+  // When the refresh buffer is hit, we trigger this refresher
+  // which tries to update the token every so often.
+  // The delay of the refreshes is defined by the `refreshAttemptsBufferMs`
+  async function refresher(tokenOptions: GetTokenOptions): Promise<AccessToken | null> {
+    // If the cached token expired, we should stop refreshing the token.
+    while (!tokenCache.isExpired()) {
+      const token = await credential.getToken(scopes, tokenOptions);
+      if (token) {
+        return token;
+      }
+      await delay(refreshAttemptsBufferMs);
+    }
+    return null;
+  }
+
   async function getToken(tokenOptions: GetTokenOptions): Promise<string | undefined> {
+    if (tokenCache.readyToRefresh()) {
+      refreshPromise = refresher(tokenOptions);
+    }
     let accessToken = tokenCache.getCachedToken();
     if (accessToken === undefined) {
-      accessToken = (await credential.getToken(scopes, tokenOptions)) || undefined;
+      if (!refreshPromise) {
+        refreshPromise = credential.getToken(scopes, tokenOptions);
+      }
+      accessToken = (await refreshPromise) || undefined;
```

Comment:
refreshPromise returns `string | null`, so now accessToken can be `string | undefined | null` ? Wouldn't that break things?

Reply:
it will be the same as before! string | undefined. Why would it be null? I didn't intend it to be null. If `await refreshPromsie` is null, it will turn unto `undefined`. Let me know if I'm missing something.

Reply:
OH of course the `||` will coalesce make it default to undefined, but why not have `refreshPromise` return undefined instead of null in the fallback case? Then you can avoid writing `|| undefined`, right?

Reply:
Can do!